libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+
+done:
+    return ret;
+}
+
+static cl_error_t findFileIdentifiers(const uint8_t *const input, PointerList *pfil)
+{
+
+    cl_error_t ret        = CL_SUCCESS;
+    const uint8_t *buffer = input;
+    uint16_t tagId        = getDescriptorTagId(buffer);
+
+    while (257 == tagId) {
+        if (CL_SUCCESS != (ret = insertPointer(pfil, buffer))) {
+            goto done;
+        }
+
+        buffer = buffer + getFileIdentifierDescriptorSize((FileIdentifierDescriptor *)buffer);
+        tagId  = getDescriptorTagId(buffer);
+    }
+
+done:
+    return ret;
+}
+
+static cl_error_t findFileEntries(const uint8_t *const input, PointerList *pfil)
+{
+
+    cl_error_t ret        = CL_SUCCESS;
+    const uint8_t *buffer = input;
+    uint16_t tagId        = getDescriptorTagId(buffer);
+
+    while (261 == tagId) {
+        if (CL_SUCCESS != (ret = insertPointer(pfil, buffer))) {
+            goto done;
+        }
+
+        buffer = buffer + getFileEntryDescriptorSize((FileEntryDescriptor *)buffer);
+        tagId  = getDescriptorTagId(buffer);
+    }
+
+done:
+    return ret;
+}
+
+cl_error_t cli_scanudf(cli_ctx *ctx, const size_t offset)
+{
+    cl_error_t ret                          = CL_SUCCESS;
+    size_t idx                              = offset;
+    size_t lastOffset                       = 0;
+    size_t i                                = 0;
+    uint8_t *buffer                         = NULL;
+    PrimaryVolumeDescriptor *pvd            = NULL;
+    GenericVolumeStructureDescriptor *gvsd  = NULL;
+    ImplementationUseVolumeDescriptor *iuvd = NULL;
+    LogicalVolumeDescriptor *lvd            = NULL;
+    PartitionDescriptor *pd                 = NULL;
+    UnallocatedSpaceDescriptor *usd         = NULL;
+    TerminatingDescriptor *td               = NULL;
+    LogicalVolumeIntegrityDescriptor *lvid  = NULL;
+    AnchorVolumeDescriptorPointer *avdp     = NULL;
+
+    bool isInitialized = false;
+    PointerList fileIdentifierList;
+    PointerList fileEntryList;
+
+    if (offset < 32768) {
+        return CL_SUCCESS; /* Need 16 sectors at least 2048 bytes long */
+    }
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, NUM_GENERIC_VOLUME_DESCRIPTORS * VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        ret = CL_SUCCESS;
+        goto done;
+    }
+
+    for (i = 0; i < NUM_GENERIC_VOLUME_DESCRIPTORS; i++) {
+        gvsd       = (GenericVolumeStructureDescriptor *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+        lastOffset = idx;
+
+        if (strncmp("BEA01", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("BOOT2", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("CD001", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("CDW02", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("NSR02", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("NSR03", gvsd->standardIdentifier, 5)) {
+        } else if (strncmp("TEA01", gvsd->standardIdentifier, 5)) {
```

Comment:
Might be nice to put some debug messages in here

Reply:
done